cdap-data-fabric/src/main/java/co/cask/cdap/data2/sql/PostgresSqlStructuredTable.java

```diff
@@ -0,0 +1,361 @@
+          statement.setInt(parameterIndex, (int) value);
+        }
+        break;
+      case LONG:
+        if (value == null) {
+          statement.setNull(parameterIndex, Types.BIGINT);
+        } else {
+          statement.setLong(parameterIndex, (long) value);
+        }
+        break;
+      case FLOAT:
+        if (value == null) {
+          statement.setNull(parameterIndex, Types.FLOAT);
+        } else {
+          statement.setFloat(parameterIndex, (float) value);
+        }
+        break;
+      case DOUBLE:
+        if (value == null) {
+          statement.setNull(parameterIndex, Types.DOUBLE);
+        } else {
+          statement.setDouble(parameterIndex, (double) value);
+        }
+        break;
+      case STRING:
+        if (value == null) {
+          statement.setNull(parameterIndex, Types.VARCHAR);
+        } else {
+          statement.setString(parameterIndex, (String) value);
+        }
+        break;
+      default:
+        // this should not happen since we validate the field before setting
+        throw new InvalidFieldException(tableSchema.getTableId(), field.getName());
+    }
+  }
+
+  private String getWriteSqlQuery(Collection<Field<?>> fields) {
+    StringJoiner insertPart = new StringJoiner(",",
+                                               "INSERT INTO " + tableSchema.getTableId().getName() + " (",
+                                               ") ");
+    StringJoiner valuePart = new StringJoiner(",", "VALUES (", ") ");
+    StringJoiner conflictPart = new StringJoiner(",", "ON CONFLICT (", ") ");
+    StringJoiner updatePart = new StringJoiner(",",  "DO UPDATE SET ", ";");
+
+    for (Field<?> field : fields) {
+      insertPart.add(field.getName());
+      valuePart.add("?");
+      if (tableSchema.isPrimaryKeyColumn(field.getName())) {
+        conflictPart.add(field.getName());
+      } else {
+        updatePart.add(field.getName() + "=EXCLUDED." + field.getName());
+      }
+    }
+    return insertPart.toString() + valuePart.toString() + conflictPart.toString() + updatePart.toString();
+  }
+
+  private String getReadQuery(Collection<Field<?>> keys, Collection<String> columns) {
+    String prefix = "SELECT " + (columns == null || columns.isEmpty() ? "*" : Joiner.on(",").join(columns)) +
+      " FROM " + tableSchema.getTableId().getName() + " WHERE ";
+    return combineWithEqualClause(prefix, keys);
+  }
+
+  private String getScanQuery(Range range, int limit) {
+    StringBuilder prefix = new StringBuilder("SELECT * FROM ").append(tableSchema.getTableId().getName());
+    if (range.getBegin() == null && range.getEnd() == null) {
+      return prefix.append(";").toString();
+    }
+    prefix.append(" WHERE ");
+    StringJoiner joiner = new StringJoiner(" AND ", prefix, " LIMIT " + limit + ";");
+    if (range.getBegin() != null) {
+      for (Field<?> key : range.getBegin()) {
+        joiner.add(key.getName() + (range.getBeginBound().equals(Range.Bound.INCLUSIVE) ? ">=?" : ">?"));
```

Comment:
is this correct? For example, if the key is (a=1,b=2), then:

- inclusive should be "WHERE a>1 OR a=1 && b>=2" because it should include (a=1,b=2) as well as (a=2,b=1)
- exclusive should be "WHERE a>1 OR a=1 AND b>2", because it should include (a=1,b=3) as well as (a=2,b=1)

Or am I misunderstanding it?

Reply:
Great catch, I have fixed the scan to work like this

Reply:
nice